gadopt/level_set_tools.py

```diff
@@ -274,258 +458,199 @@ def gradient_L2_proj(self) -> fd.variational_solver.LinearVariationalSolver:
+                reinitialisation_equation,
+                self.solution,
+                self.reini_kwargs["timestep"],
+                solution_old=self.solution_old,
+                solver_parameters=self.reini_kwargs["solver_params"],
+            )
+
+        self.step = 0
+        self._solvers_ready = True
+
+    def update_level_set_gradient(self) -> None:
         """Calls the gradient projection solver.
 
-        Can be used as a callback.
+        Can be provided as a callback to time integrators.
         """
         self.proj_solver.solve()
 
-    def set_up_solvers(self):
-        """Sets up the time steppers for advection and reinitialisation."""
-        test = fd.TestFunction(self.func_space)
-
-        self.ls_solver = GenericTransportSolver(
-            "advection",
-            self.level_set,
-            self.tstep / self.subcycles,
-            self.tstep_alg,
-            eq_attrs={"u": self.u},
-            solver_parameters=self.solver_params,
-        )
-
-        reinitialisation_equation = Equation(
-            test,
-            self.func_space,
-            reinitialisation_term,
-            mass_term=scalar_eq.mass_term,
-            eq_attrs=self.reini_eq_attrs,
-        )
-        self.reini_ts = self.reini_params["tstep_alg"](
-            reinitialisation_equation,
-            self.level_set,
-            self.reini_params["tstep"],
-            solver_parameters=self.solver_params,
-        )
-
-    def solve(self, step: int):
-        """Updates the level-set function.
+    def reinitialise(self) -> None:
+        """Performs reinitialisation steps."""
+        for _ in range(self.reini_kwargs["steps"]):
+            self.reini_integrator.advance(
+                update_forcings=self.update_level_set_gradient
+            )
 
-        Calls advection and reinitialisation solvers within a subcycling loop.
-        The reinitialisation solver can be iterated and might not be called at each
-        simulation time step.
+    def solve(
+        self,
+        disable_advection: bool = False,
+        disable_reinitialisation: bool = False,
+    ) -> None:
+        """Updates the level-set function by means of advection and reinitialisation.
 
         Args:
-            step:
-              An integer representing the current simulation step.
+          disable_advection:
+            A boolean to disable the advection solve.
+          disable_reinitialisation:
+            A boolean to disable the reinitialisation solve.
         """
-        if not self.solvers_ready:
+        if not self._solvers_ready:
             self.set_up_solvers()
-            self.solvers_ready = True
-
-        for subcycle in range(self.subcycles):
-            self.ls_solver.solve()
 
-            if step >= self.reini_params["frequency"]:
-                self.reini_ts.solution_old.assign(self.level_set)
+        if self.advection and not disable_advection:
```

Comment:
Do you need two ways of turning on/off advection/reinitialisation?

Reply:
`self.advection` internally records if the advection part of the solver has been initialised, whereas `disable_advection` is provided at solve time to indicate if advection should be omitted for the particular solve to follow. Essentially, I would say there is only one way of turning advection or reinitialisation off, but given that I use optional booleans and give them a default value of `True`, I need a way to make sure that a solver only executes if it has been initialised.

Reply:
Fair enough!